providers/providers.go

```diff
@@ -162,6 +162,7 @@ func newProviderDataFromConfig(providerConfig options.Provider) (*ProviderData,
 	if providerConfig.OIDCConfig.UserIDClaim == "" {
 		providerConfig.OIDCConfig.UserIDClaim = "email"
 	}
+	p.UserClaim = providerConfig.OIDCConfig.UserIDClaim
```

Comment:
Going back over this, I don't think this is correct. UserID != User

UserID used to be grabbing the email, while the email was already being grabbed, if we look at the current code, UserID is the `sub`, which is not currently user configurable it claims

https://github.com/oauth2-proxy/oauth2-proxy/blob/ce750e9b301ac66b3b8cbdbd0bc8ac7415bfed79/providers/provider_data.go#L23-L24

Looking through the history, we have never had anything setting the user claim

Reply:
Thanks for the explanation and correcting my misinterpretation of the `OIDCConfig.UserIDClaim` setting.

The root of the issue is that the provider `UserClaim` is not set. This is needed when building the session here: 
https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/provider_data.go#L255

I noticed this issue when we upgraded from v7.0.1 where we can see that it used to set this in the session based on the subject:
https://github.com/oauth2-proxy/oauth2-proxy/blob/v7.0.1/providers/provider_data.go#L119-L154

Therefore with the understanding above I believe we should be initializing `UserClaim` to `sub` if it is not set.
```suggestion
	if p.UserClaim == "" {
		p.UserClaim = oidcUserClaim
	}
```

Let me know if this approach makes sense and I will update the PR accordingly.

Reply:
I think we already have that covered https://github.com/oauth2-proxy/oauth2-proxy/blob/ce750e9b301ac66b3b8cbdbd0bc8ac7415bfed79/providers/provider_data.go#L201-L203

Or at least we should :thinking:

Reply:
The issue is that not every provider calls the `setProviderDefaults` function.

For more context, you can see this regression was introduced in the following PR: 
https://github.com/oauth2-proxy/oauth2-proxy/commit/967051314e5d84cb389aa33ca8e0ff624b43c7eb#diff-b4ef38e96ab5394c070aff6c070843820d904f7f06b36a81bd52384edd5214caR107-L159
Previously the user in the session was to set the `sub` claim for all providers.

One possible solution is to update all providers to ensure each invokes `setProviderDefaults`.

However, I think it makes more sense to move the setting of `UserClaim` to the `newProviderDataFromConfig` function because it is responsible for initializing the provider data and `UserClaim` is not a value that can be overwritten in the `setProviderDefaults` function.
https://github.com/oauth2-proxy/oauth2-proxy/blob/ce750e9b301ac66b3b8cbdbd0bc8ac7415bfed79/providers/provider_data.go#L181-L204